// @azure/storage-file-datalake/dist-esm/src/generated/src/operations/service.js
/*
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License. See License.txt in the project root for
 * license information.
 *
 * Code generated by Microsoft (R) AutoRest Code Generator.
 * Changes may cause incorrect behavior and will be lost if the code is
 * regenerated.
 */
import * as coreHttp from "@azure/core-http";
import * as Mappers from "../models/serviceMappers";
import * as Parameters from "../models/parameters";
/** Class representing a Service. */
var Service = /** @class */ (function () {
    /**
     * Create a Service.
     * @param {StorageClientContext} client Reference to the service client.
     */
    function Service(client) {
        this.client = client;
    }
    Service.prototype.listFileSystems = function (options, callback) {
        return this.client.sendOperationRequest({
            options: options
        }, listFileSystemsOperationSpec, callback);
    };
    return Service;
}());
export { Service };
// Operation Specifications
var serializer = new coreHttp.Serializer(Mappers);
var listFileSystemsOperationSpec = {
    httpMethod: "GET",
    urlParameters: [
        Parameters.url
    ],
    queryParameters: [
        Parameters.resource0,
        Parameters.prefix,
        Parameters.continuation,
        Parameters.maxResults,
        Parameters.timeout
    ],
    headerParameters: [
        Parameters.requestId,
        Parameters.version
    ],
    responses: {
        200: {
            bodyMapper: Mappers.FileSystemList,
            headersMapper: Mappers.ServiceListFileSystemsHeaders
        },
        default: {
            bodyMapper: Mappers.StorageError,
            headersMapper: Mappers.ServiceListFileSystemsHeaders
        }
    },
    serializer: serializer
};
